components/eam/src/physics/cam/shoc.F90

```diff
@@ -3413,25 +3413,19 @@ subroutine vd_shoc_decomp( &
   du(:,nlev) = 0._rtype
   dl(:,1)    = 0._rtype
 
-  ! The diagonal elements are a combination of du and dl (d=1-du-dl). Surface
-  ! fluxes are applied explicitly in the diagonal at the top level.
-  do i=1,shcol
-    d(i,nlev) = 1._rtype - dl(i,nlev) + flux(i)*dtime*ggr*rdp_zt(i,nlev)
-  enddo
-  do k=nlev-1,2,-1
+  do k=1,nlev
     do i=1,shcol
+      ! Compute the diagonal. The diagonal elements are a combination
+      ! of du and dl (d=1-du-dl). Surface fluxes are applied explicitly
+      ! in the diagonal at the top level.
       d(i,k) = 1._rtype - du(i,k) - dl(i,k)
-    enddo
-  enddo
-  do i=1,shcol
-    d(i,1) = 1._rtype - du(i,1)
-  enddo
+      if (k .eq. nlev) d(i,k) = d(i,k) + flux(i)*dtime*ggr*rdp_zt(i,nlev)
 
-  ! Compute the Thomas factorization
-  do k=2,nlev
-    do i=1,shcol
-      dl(i,k) = dl(i,k)/d(i,k-1)
-      d (i,k) = d (i,k) - dl(i,k)*du(i,k-1)
+      ! Compute the Thomas factorization
+      if (k .ne. 1) then
```

Comment:
Sorry to be picky, but we really shouldn't have `if` blocks inside the `i` loop. I recommend just moving the `k=1` case outside the `k` loop. A less attractive alternative, but not wholly unjustified, is to move the `i` loop into the `if` blocks.

Reply:
If we remove both if-statements out of the i-loop then I would need the factorization code twice:
```
do i=1,shcol
  d(i,1) = 1._rtype - du(i,1)
enddo
do k=2,nlev-1
  do i=1,shcol
    d(i,k) = 1._rtype - du(i,k) - dl(i,k)

    dl(i,k) = dl(i,k)/d(i,k-1)
    d (i,k) = d (i,k) - dl(i,k)*du(i,k-1)
  enddo
enddo
do i=1,shcol
  d(i,nlev) = 1._rtype - dl(i,nlev) + flux(i)*dtime*ggr*rdp_zt(i,nlev)

  dl(i,nlev) = dl(i,nlev)/d(i,nlev-1)
  d (i,nlev) = d (i,nlev) - dl(i,nlev)*du(i,nlev-1)
enddo
```

The code is _slightly_ more concise if we separate the diagonal and factorization computation:
```
do k=1,nlev
  do i=1,shcol
    d(i,k) = 1._rtype - du(i,k) - dl(i,k)
  enddo
enddo
do i=1,shcol
  d(i,nlev) = d(i,nlev) + flux(i)*dtime*ggr*rdp_zt(i,nlev)
enddo

do k=2,nlev
  do i=1,shcol
    dl(i,k) = dl(i,k)/d(i,k-1)
    d (i,k) = d (i,k) - dl(i,k)*du(i,k-1)
  enddo
enddo
```
What's your opinions? I'm ok with either.

Reply:
I'd say either use the first above or revert to two loops. I personally like the first one above. It's pretty standard to have to deal with boundary conditions, and they do tend to lead to slightly more verbose code if the code is written to cache well, move through memory linearly, and vectorize well.

Reply:
I went with the first above.